python/ray/_private/worker.py

```diff
@@ -829,9 +831,32 @@ def put_object(
                 "If you really want to do this, you can wrap the "
                 "ray.ObjectRef in a list and call 'put' on it."
             )
-
+        tensors = None
+        tensor_transport = TensorTransportEnum.from_str(tensor_transport)
+        tensor_transport_meta = None
         try:
-            serialized_value = self.get_serialization_context().serialize(value)
+            if tensor_transport != TensorTransportEnum.OBJECT_STORE:
+                from ray.experimental.collective import get_tensor_transport_manager
+                from ray.experimental.gpu_object_manager.gpu_object_store import (
+                    _tensor_transport_to_collective_backend,
+                )
+
+                (
+                    serialized_value,
+                    tensors,
+                ) = self.get_serialization_context().serialize_gpu_objects(value)
+                tensor_transport_backend = _tensor_transport_to_collective_backend(
+                    tensor_transport
+                )
+                if tensors:
+                    transport_manager = get_tensor_transport_manager(
+                        tensor_transport_backend
+                    )
+                    tensor_transport_meta = (
+                        transport_manager.extract_tensor_transport_metadata(tensors)
+                    )
```

Comment:
For readability and to keep the GPU object code more self-contained, I think it'd be better to move this logic into a single `gpu_object_manager.put` call, along with the logic [here](https://github.com/ray-project/ray/pull/56146/files#diff-bec668cc316d34704cdc3e37dd4ce18e4e87764fdf2c28b1504478c7eb8291deR890-R897).